bellman/src/domain/fft.rs

```diff
@@ -0,0 +1,177 @@
+use crate::{multicore, arith};
+use multicore::Worker;
+use arith::Group;
+
+use ff::{Field, ScalarEngine};
+
+pub fn run_optimal_fft<E,T>(a: &mut [T], worker: &Worker, omega: &E::Fr, log_n: u32) 
+where
+    E: ScalarEngine,
+    for <'a> T: Group<'a,E> 
+{
+    let log_cpus = worker.log_num_cpus();
+
+    if log_n <= log_cpus {
+        serial_fft(a, omega, log_n);
+    } else {
+        parallel_fft(a, worker, omega, log_n, log_cpus);
+    }
+}
+
+pub fn serial_fft<E,T>(series: &mut [T], omega: &E::Fr, log_n: u32) 
+where 
+    E: ScalarEngine, 
+    for <'a> T: Group<'a,E>
+{
+    let len: u32 = series.len() as u32;
+    assert_eq!(len, 1 << log_n);
+
+    for k in 0..len {
+        let rk: u32 = bitreverse(k, log_n);
```

Comment:
More meaningful names if possible.

Reply:
I'm fairly stumped on naming for the fft implementation at the moment. If you have any suggestions I'm happy to put them in. This all looks like pseudocode implementations that haven't been given a source :/